cinder/scheduler/manager.py

```diff
@@ -471,15 +471,29 @@ def _extend_volume_set_error(self, context, ex, request_spec):
                 {'size': new_size - volume.size})
             volume_rpcapi.VolumeAPI().extend_volume(context, volume, new_size,
                                                     reservations)
-        except exception.NoValidBackend as ex:
-            QUOTAS.rollback(context, reservations,
-                            project_id=volume.project_id)
-            _extend_volume_set_error(self, context, ex, request_spec)
-            self.message_api.create(
-                context,
-                message_field.Action.EXTEND_VOLUME,
-                resource_uuid=volume.id,
-                exception=ex)
+        except exception.NoValidBackend:
+            try:
+                self._extend_migrate(context, volume, new_size, request_spec,
+                                     filter_properties, reservations)
+            except exception.NoValidBackend as ex:
+                QUOTAS.rollback(context, reservations,
+                                project_id=volume.project_id)
+                _extend_volume_set_error(self, context, ex, request_spec)
+                self.message_api.create(
+                    context,
+                    message_field.Action.EXTEND_VOLUME,
+                    resource_uuid=volume.id,
+                    exception=ex)
+
+    def _extend_migrate(self, context, volume, new_size, request_spec,
+                        filter_properties, reservations):
+        backend_state = self.driver.find_backend_for_extend(
+            context, request_spec, filter_properties)
+
+        volume_rpcapi.VolumeAPI().migrate_volume(
+            context, volume, backend_state,
+            force_host_copy=False, wait_for_completion=False,
+            extend_spec={'new_size': new_size, 'reservations': reservations})
```

Comment:
Why do we add extending the volume to migrating the volume? Is this, because `migrate_volume` is a cast and we don't know when it's done and thus we don't know when to call extend on it?

Reply:
Yes, migrate is a cast and the extend is triggered by the volume service, once the migration ends.
This way we avoid any timeouts given by slow migrations.